android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java

```diff
@@ -0,0 +1,489 @@
+import android.security.keystore.KeyGenParameterSpec;
+import android.security.keystore.KeyInfo;
+import android.text.TextUtils;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.oblador.keychain.SecurityLevel;
+import com.oblador.keychain.exceptions.CryptoFailedException;
+import com.oblador.keychain.exceptions.KeyStoreAccessException;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.ProviderException;
+import java.security.UnrecoverableKeyException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.spec.IvParameterSpec;
+
+import static com.oblador.keychain.SecurityLevel.SECURE_HARDWARE;
+
+@SuppressWarnings({"unused", "WeakerAccess"})
+abstract public class CipherStorageBase implements CipherStorage {
+  //region Constants
+  /** Logging tag. */
+  protected static final String LOG_TAG = CipherStorageBase.class.getSimpleName();
+  /** Default key storage type/name. */
+  public static final String KEYSTORE_TYPE = "AndroidKeyStore";
+  /** Key used for testing storage capabilities. */
+  public static final String TEST_KEY_ALIAS = KEYSTORE_TYPE + "#supportsSecureHardware";
+  /** Default service name. */
+  public static final String DEFAULT_ALIAS = "RN_KEYCHAIN_DEFAULT_ALIAS";
+  /** Size of hash calculation buffer. Default: 4Kb. */
+  private final static int BUFFER_SIZE = 4 * 1024;
+  /** Default size of read/write operation buffer. Default: 16Kb. */
+  private final static int BUFFER_READ_WRITE_SIZE = 4 * BUFFER_SIZE;
+  /** Default charset encoding. */
+  public static final Charset UTF8 = Charset.forName("UTF-8");
+  //endregion
+
+  //region Overrides
+
+  /** Hardware supports keystore operations. */
+  @Override
+  public SecurityLevel securityLevel() {
+    return SecurityLevel.SECURE_HARDWARE;
+  }
+
+  /**
+   * The higher value means better capabilities. Range: [19..1129].
+   * Formula: `1000 * isBiometrySupported() + 100 * isSecureHardware() + minSupportedApiLevel()`
+   */
+  @Override
+  public final int getCapabilityLevel() {
+    // max: 1000 + 100 + 29 == 1129
+    // min: 0000 + 000 + 19 == 0019
+
+    return
+      (1000 * (isBiometrySupported() ? 1 : 0)) + // 0..1000
+        (100 * (supportsSecureHardware() ? 1 : 0)) + // 0..100
+        (getMinSupportedApiLevel()); // 19..29
+  }
+
+  /** Try device capabilities by creating temporary key in keystore. */
+  @Override
+  public boolean supportsSecureHardware() {
+    try (SelfDestroyKey sdk = new SelfDestroyKey(TEST_KEY_ALIAS)) {
+      return validateKeySecurityLevel(SECURE_HARDWARE, sdk.key);
```

Comment:
To follow code style used in this file:
```suggestion
      return validateKeySecurityLevel(SecurityLevel.SECURE_HARDWARE, sdk.key);
```